bug: revision doesn't transition from root to child pom files

I've been getting this bug, which I think happens because, somehow, the property `revision` doesn't transition from parent to child pom files.

```
[ERROR] Failed to execute goal fr.inria.gforge.spirals:repair-maven-plugin:1.6-SNAPSHOT:npefix (default) on project commons-collection: Execution default of goal fr.inria.gforge.spirals:repair-maven-plugin:1.6-SNAPSHOT:npefix failed: Plugin fr.inria.gforge.spirals:repair-maven-plugin:1.6-SNAPSHOT or one of its dependencies could not be resolved: Failed to collect dependencies at fr.inria.gforge.spirals:repair-maven-plugin:jar:1.6-SNAPSHOT -> fr.inria.repairnator:repairnator-core:jar:3.3-SNAPSHOT: Failed to read artifact descriptor for fr.inria.repairnator:repairnator-core:jar:3.3-SNAPSHOT: Failure to find fr.inria.repairnator:repairnator-root:pom:${revision} in https://repo.maven.apache.org/maven2 was cached in the local repository, resolution will not be reattempted until the update interval of central has elapsed or updates are forced -> [Help 1]
```

This concrete error is from testing `maven-repair` on lang304. To reproduce: 
```
❯ cd repairnator/
❯ mvn clean install -DskipTests
❯ cd npe-dataset/lang-304/
❯ mvn clean test -DtrimStackTrace=false
❯ mvn fr.inria.gforge.spirals:repair-maven-plugin:1.6-SNAPSHOT:npefix
```
I've been hard-coding the version and replacing `${revision}` as a work-around, but this should not be necessary.



Now comes the interesting part. @cokefh wasn't able to reproduce this bug (nor the one from #1153 ), while both in my laptop and desktop they occur (almost exact environment between both).

I've tried deleting my local `.m2` and using a clean git project, but nothing changed. I also didn't find anything online mentioning this bug.

Information about my desktop setup:
```
❯ uname -a
Linux alan 5.9.11-arch2-1 #1 SMP PREEMPT Sat, 28 Nov 2020 02:07:22 +0000 x86_64 GNU/Linux
```
```
❯ java -version
java version "1.8.0_271"
Java(TM) SE Runtime Environment (build 1.8.0_271-b09)
Java HotSpot(TM) 64-Bit Server VM (build 25.271-b09, mixed mode)
```
```
❯ mvn -version
Apache Maven 3.6.3 (NON-CANONICAL_2019-11-27T20:26:29Z_root)
Maven home: /opt/maven
Java version: 1.8.0_271, vendor: Oracle Corporation, runtime: /usr/lib/jvm/java-8-jdk/jre
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "5.9.11-arch2-1", arch: "amd64", family: "unix"
```

--- a/.ci/ci-deploy.sh
+++ b/.ci/ci-deploy.sh
@@ -6,8 +6,6 @@
 set -e
 
 sudo apt-get update
-sudo apt-get install -y xmlstarlet
-
 
 ### MAVEN CENTRAL
 if [ "$TRAVIS_PULL_REQUEST" = "false" ] && [ "$TRAVIS_BRANCH" = "master" ]; then
@@ -15,9 +13,6 @@ if [ "$TRAVIS_PULL_REQUEST" = "false" ] && [ "$TRAVIS_BRANCH" = "master" ]; then
     cd src 
     
     # pushing snapshot to https://oss.sonatype.org/content/repositories/snapshots/fr/inria/repairnator/
-    VERSION=`xmlstarlet sel -t -v '//_:project/_:properties/_:revision' pom.xml`
-    sed -i -e 's/\${revision}/'$VERSION'/' pom.xml */pom.xml
-    git diff
     mvn deploy -DskipTests -Dcheckstyle.skip
     echo Deployment to Maven Central done
 fi

--- a/.ci/ci-dockerfile
+++ b/.ci/ci-dockerfile
@@ -1,3 +1,4 @@
 FROM maven:3.6-openjdk-8
 
 RUN apt-get update
+RUN apt-get install xmlstarlet -y

--- a/.ci/ci-dockerfile
+++ b/.ci/ci-dockerfile
@@ -1,4 +1,3 @@
 FROM maven:3.6-openjdk-8
 
 RUN apt-get update
-RUN apt-get install xmlstarlet -y

--- a/.ci/ci-maven-repair.sh
+++ b/.ci/ci-maven-repair.sh
@@ -4,11 +4,6 @@
 set -e
 export M2_HOME=/usr/local/maven
 
-VERSION=`xmlstarlet sel -t -v '//_:project/_:properties/_:revision' src/pom.xml`
-sed -i -e 's/\${revision}/'$VERSION'/' src/repairnator-core/pom.xml
-sed -i -e 's/\${revision}/'$VERSION'/' src/pom.xml
-sed -i -e 's/\${revision}/'$VERSION'/' "${TEST_PATH}"/pom.xml
-
 NPEFIX_VERSION=`xmlstarlet sel -t -v '//_:dependency[_:artifactId="npefix"]/_:version' src/maven-repair/pom.xml`
 
 mvn clean install -B -f src/repairnator-core/ && mvn -Dtest=$TEST_LIST -DNPEFIX_VERSION=$NPEFIX_VERSION clean test -B -f $TEST_PATH -DskipTests

--- a/.ci/ci-run-with-core.sh
+++ b/.ci/ci-run-with-core.sh
@@ -8,9 +8,4 @@ export M2_HOME=/usr/local/maven
 # sudo apt-get update
 # sudo apt-get install -y xmlstarlet
 
-VERSION=`xmlstarlet sel -t -v '//_:project/_:properties/_:revision' src/pom.xml`
-sed -i -e 's/\${revision}/'$VERSION'/' src/repairnator-core/pom.xml
-sed -i -e 's/\${revision}/'$VERSION'/' src/pom.xml
-sed -i -e 's/\${revision}/'$VERSION'/' "${TEST_PATH}"/pom.xml
-
 mvn clean install -B -f src/repairnator-core/ && mvn -Dtest=$TEST_LIST clean test -B -f $TEST_PATH -DskipTests

--- a/.gitignore
+++ b/.gitignore
@@ -5,6 +5,7 @@ pom.xml.tag
 pom.xml.releaseBackup
 pom.xml.versionsBackup
 pom.xml.next
+.flattened-pom.xml
 release.properties
 dependency-reduced-pom.xml
 buildNumber.properties

--- a/src/pom.xml
+++ b/src/pom.xml
@@ -116,6 +116,31 @@
     <build>
         <plugins>
             <plugin>
+                <groupId>org.codehaus.mojo</groupId>
+                <artifactId>flatten-maven-plugin</artifactId>
+                <version>1.1.0</version>
+                <configuration>
+                    <updatePomFile>true</updatePomFile>
+                    <flattenMode>resolveCiFriendliesOnly</flattenMode>
+                </configuration>
+                <executions>
+                    <execution>
+                        <id>flatten</id>
+                        <phase>process-resources</phase>
+                        <goals>
+                            <goal>flatten</goal>
+                        </goals>
+                    </execution>
+                    <execution>
+                        <id>flatten.clean</id>
+                        <phase>clean</phase>
+                        <goals>
+                            <goal>clean</goal>
+                        </goals>
+                    </execution>
+                </executions>
+            </plugin>
+            <plugin>
                 <groupId>org.jacoco</groupId>
                 <artifactId>jacoco-maven-plugin</artifactId>
                 <version>${maven-jacoco-plugin.version}</version>

